src/wasm/wasm-type.cpp

```diff
@@ -1436,42 +1989,46 @@ Canonicalizer::Canonicalizer(TypeBuilder& builder) : builder(builder) {
     }
   }
 
-  findSelfReferentialHeapTypes();
-
-  // Visit the use sites of Types and HeapTypes, replacing them with their
-  // canonicalized versions in an order that guarantees no temporary types will
-  // be leaked into the global stores.
-  for (auto it : getOrderedItems()) {
-    switch (it.kind) {
-      case Item::TypeKind:
-        canonicalize(it.type, canonicalTypes);
-        break;
-      case Item::HeapTypeKind:
-        canonicalize(it.heapType, canonicalHeapTypes);
-        break;
+  // Canonicalize HeapTypes at all their use sites. HeapTypes for which there
+  // was not already a globally canonical version are moved to the global store
+  // to become the canonical version. These new canonical HeapTypes still
+  // contain references to temporary Types owned by the TypeBuilder, so we must
+  // subsequently replace those references with references to canonical Types.
+  // Canonicalize non-tuple types before tuple types to avoid canonicalizing a
+  // tuple that still contains non-canonical Types.
```

Comment:
Why are tuples special? Structs also contain a list of children just like a Tuple so the difficulty seems the same?

Reply:
Tuples are the only types that directly contain other types. Structs are HeapTypes that contain types, but we've already canonicalized HeapTypes.

Reply:
Oh - that is rather surprising that such a difference causes a need for special handling here, but it seems unavoidable.

Reply:
This used to be handled by performing a topological sort on all the types and heap types, but now that none of the heap types need to participate in that topological sort, the sort seemed like overkill compared to just deferring the canonicalization of tuples.